### ISSUE TITLE
Contribution guidelines: files and paths

### ISSUE THREAD

**Issue by rprieto** [tldr-pages/tldr#478]
It would be nice to come up with a standard across TLDR to have a consistent approach to `{{file}}`, `{{file.ext}}`, `{{path/to/file}}`, `{{/path/dir/file}}`, {{path_to_file}}`.

Personally I'd say `{{file}}` is enough for  80% of cases, and is generic enough that anyone with basic Unix knowledge knows you can specify a file name, relative path, absolute path.... or even the `<(command)` operator in bash. I can't think of a command that would specifically **need** a relative or absolute path.

We'd also need guidelines for acceptable exceptions, for example
- specifying the extension is acceptable if it matters to the command, e.g. `open {{file.ext}}`
- using other words like `source` or `target` if it clarifies the example, e.g. `tar cfz {{target.tar.gz}}`
- using `file1`, `file2`... if there are multiple files in the command, e.g. `tar cf {{target.tar}} {{file1}} {{file2}}`

**Comment by igorshubovych** [comment]
Cc @rprieto

**Comment by rprieto** [comment]
Why the cc?

**Comment by waldyrious** [comment]
I mostly agree w/ @rprieto, but I'd add that making the paths explicit is useful at least for commands dealing with two locations, e.g, `cp`, `mv`, `scp`, etc, or with remote locations.

**Comment by igorshubovych** [comment]
oh, sorry Romain. i thought somebody else opened it :)

**Comment by rubenvereecken** [comment]
I wouldn't mind writing a draft is no one else is eager to, and then we can iterate on that. I followed most of the discussions related to this one so I've got a fairly recent overview, now it's about wording it concisely.

**Comment by igorshubovych** [comment]
I would add that path can be used for specific folders:

``virtualenv {{path/to/venv}}``
`usermod -m -d {{/path/to/home}} {{user}}`

**Comment by rprieto** [comment]
Very good point differentiating files/folders. To avoid the questions about absolute/relative, whether to start with a `/` etc... could we use `{{file}}` and `{{dir}}`, or sometimes being more specific like `{{home_dir}}`? It also allows us to make it very clear when a command needs a trailing slash, e.g. `somecommand {{source_dir}}/`

**Comment by waldyrious** [comment]
What about commands that take either files or directories? We've discussed about the (ugly IMO) `{{file_or_dir}}` somewhere recently, and decided against it IIRC.

**Comment by agnivade** [comment]
I recently made a [commit](https://github.com/tldr-pages/tldr/commit/fa8b2d8f92abfcbea46036b8a30c129ac53abdcb) which standardised the usage of `path/to/item` convention throughout the repo. I am jotting down the points that were followed and also some additional guidelines which should standardise all references to files, paths and extensions. 

**File / Folder stuff**
- Any path related reference to file or folder must be mentioned like this.
eg:
{{path/to/file}}
{{path/to/folder}}

- If something else other than just `file` or `folder`, better clarifies your example use that.
eg:
{{path/to/symlink}}
{{path/to/executable}}

- If the placeholder has multiple words, apply snake case.
eg: 
{{path/to/iptables_file}}
{{path/to/raid_device_file}}

- For the case of reference to a file or a folder, use this.
{{path/to/file/or/folder}}

**Extensions**
- If mentioning the actual extension explains the command better, use it.
eg:
{{file.png}}

- If multiple extensions are allowed , separate by | and wrap the entire thing by (). **This is not followed in places. Need to fix**
eg:
{{file.(scss|sass)}}

alternative to this - {{file.(scss,sass)}}

- If the extension is best left generic, use 'ext'.
eg:
{{file.ext}}

**Concluding points**
- If the file name contains multiple words, again follow snake case. **Not followed in places. Need to fix.**

- The file and extension rules can be combined.
eg:
{{path/to/file.apk}}

Lastly, the {{path/to}} convention is usually followed when there is a path related command. Not when the file location is implicit (eg- compiler commands like gcc,rustc or text editor commands). In that case, simply writing {{file}} is the simplest course. But of course, proper judgement is to be used, keeping simplicity and user friendliness as the top priority.

I will make a PR once the 'need to fix' items are agreed upon. And then finally we can add a section about this in the CONTRIBUTING page.

**Comment by waldyrious** [comment]
@agnivade great summary! I agree with most of it. A few observations:
- perhaps change "{{path/to/file/or/folder}}" to "{{path/to/file_or_folder}}"?
- (...|...) is a nice choice, but I'm pretty sure a syntax multiple options has been discussed previously in another thread (a PR, IIRC), so we should check that out to see if there are important arguments we might be overlooking
- "If the extension is best left generic, use 'ext'. eg: {{file.ext}}" <-- or don't use an extension at all?
- "If the file name contains multiple words, again follow snake case. Not followed in places. Need to fix." <-- huh, I thought you had handled the non-snake-case instances in #947. What kinds of variations have you found?

**Comment by waldyrious** [comment]
I think I found the thread where separators had been addressed bewfore: #719 :)

**Comment by agnivade** [comment]
@waldyrious 
- Sure. I'm good with that too.
- Saw the PR. {{file.scss|file.sass}} definitely looks much neater.
- Didn't get this point. How can we not use an extension ? We need to use "something". Maybe you meant something else ?
- Ahh not really, I just applied the 'path/to/' convention wherever a file path was mentioned. Here are some cases that I found:
```
pages/common/git-diff.md:`git diff > {{target-file.patch}}`
pages/common/gzip.md:`gzip -c {{file.ext}} > compressed-file.ext.gz`
pages/common/gzip.md:`gzip -c -d {{file.ext.gz}} > uncompressed-file.ext`
pages/common/gzip.md:`gzip -9 -c {{file.ext}} > compressed-file.ext.gz`
```
Another variation I have found is continuous non-underscored words:
```
pages/common/nc.md:`nc {{ip_address}} {{port}} > somefile.txt`
pages/common/parallel.md:`cat {{bigfile.txt}} | parallel --pipe --block 1M {{command}}`
```

**Comment by waldyrious** [comment]
> Didn't get this point. How can we not use an extension

I meant that if the extension needs to be generic, perhaps the command isn't any clearer if we include an extension in the token anyway? For instance, there are commands that accept filenames only, not paths, so the token is {{filename}}. That's taken to mean any filename, whether it includes an extension or not. This is even more relevant in Unix-like systems, where file types aren't determined by the extension. Does that make sense, or are we talking about different things?

Thanks for the clarification about the non-snake-case tokens. Indeed those need to be fixed, but that should be pretty uncontroversial (at the time those were added I believe we hadn't decided on the snake_case convention yet).

**Comment by agnivade** [comment]
Ah got you now. Yes, we were talking the same thing. My feeling is that adding the extension does sometime make sense. Take for example, the gzip page.

```
- Compress a file specifying the output filename:
`gzip -c {{file.ext}} > compressed-file.ext.gz`
```

I think specifying ".ext" here makes it a bit clearer. Although technically, just specifying {{file}} is also not wrong. But there's no harm in giving the extension.

Another one is the open page for osx

```
- Open all the files of a given extension in the current directory with the associated application:
`open {{*.ext}}`
```

Here, clearly specifying .ext makes sense. 
But I don't feel too strongly about this. If it helps clarify the command, mention it, else keep it short.


> Indeed those need to be fixed, but that should be pretty uncontroversial 

Agree. Now that we have the snake_case convention, when I send a PR, it will be an exhaustive one. Then every page will be uniform.

**Comment by waldyrious** [comment]
> If it helps clarify the command, mention it, else keep it short.

Agreed. We can say that explicitly state in the guidelines that .ext should be used if it clarifies the command, and give the examples you provide above to explain what we mean :)

**Comment by agnivade** [comment]
Perfect.

**Comment by agnivade** [comment]
@waldyrious - I will wait for your #962 to be pushed. Then I will update the CONTRIBUTING.md with these guidelines.

**Comment by waldyrious** [comment]
@agnivade sure. I've already added a proposed list of commands to the git section -- check it out and let me know what you think :)

**Comment by agnivade** [comment]
@waldyrious - can this be closed now ?

**Comment by waldyrious** [comment]
Yup. Closing per #947 and #981.

**Comment by waldyrious** [comment]
For future reference, early discussion about whether the tokens should be concrete examples or generic descriptiors happened in #27.

**Comment by waldyrious** [comment]
Also for future reference, later discussion about whether to use "folder" or "directory" (resolved in favor of the latter) happened in #2751.
